src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -152,7 +152,8 @@ G1CMMarkStack::TaskQueueEntryChunk* G1CMMarkStack::ChunkAllocator::allocate_new_
 
     MutexLocker x(MarkStackChunkList_lock, Mutex::_no_safepoint_check_flag);
     if (Atomic::load_acquire(&_buckets[bucket]) == nullptr) {
-      if (!expand()) {
+      size_t new_capacity = MIN2(bucket_size(bucket) * 2, _max_capacity);
+      if (!expand(new_capacity)) {
         return nullptr;
       }
     }
```

Comment:
As I understand it, the issue is that `expand` doesn't guarantee that `_buckets[bucket]` is filled. What do you think of the idea of using a `while` loop here? (one doesn't need a new API and the `*2` logic stays centralized.

Reply:
I prefer a thread expands to the required capacity instead of the expansion that doesn't consider the thread context.